kuwala/core/canvas/src/pages/DataSourcePreview.js

```diff
@@ -121,33 +121,118 @@ export default () => {
         });
 
         if(res.status === 200) {
-            setTableDataPreview(res.data)
+            let cols = res.data.columns.map((el,i)=>{
+                return {
+                    Header: el,
+                    accessor: el,
+                }
+            });
+
+            cols = [{
+                Header: "#",
+                id: "row",
+                filterable: false,
+                width: 50,
+                Cell: (row) => {
+                    return <div>{row.index}</div>;
+                }
+            }, ...cols]
+
+            setTableDataPreview({
+                columns: cols,
+                rows: getDataDictionary(res.data.rows, res.data.columns),
+            });
         }
         setIsTableDataPreviewLoading(false)
     }
 
+    const getDataDictionary = (data, headers) => {
+        let dictionary = [];
+
+        for (let i = 1; i < data.length; i++) {
+            let object = {};
+            for (let j = 0; j < data[i].length; j++) {
+                object[headers[j]] = typeof data[i][j] === 'object' ? JSON.stringify(data[i][j]) : data[i][j];
+            }
+            dictionary.push(object);
+        }
```

Comment:
Feels like this can be written crisper and more easily readable by using `.map()` functions.